homeassistant/helpers/script.py

```diff
@@ -121,180 +132,82 @@ class _SuspendScript(Exception):
-                if not data:
-                    continue
-
-                entity_ids = data.get(ATTR_ENTITY_ID)
+    def name(self):
+        """Return script name."""
+        return self._script.name
 
-                if entity_ids is None:
-                    continue
-
-                if isinstance(entity_ids, str):
-                    entity_ids = [entity_ids]
-
-                for entity_id in entity_ids:
-                    referenced.add(entity_id)
-
-            elif action == ACTION_CHECK_CONDITION:
-                referenced |= condition.async_extract_entities(step)
+    @property
+    def last_action(self):
+        """Return last action."""
+        return self._last_action
 
-            elif action == ACTION_ACTIVATE_SCENE:
-                referenced.add(step[CONF_SCENE])
+    @last_action.setter
+    def last_action(self, value):
+        self._last_action = value
 
-        self._referenced_entities = referenced
-        return referenced
-
-    def run(self, variables=None, context=None):
+    @abstractmethod
+    async def async_run(self) -> None:
         """Run script."""
-        asyncio.run_coroutine_threadsafe(
-            self.async_run(variables, context), self.hass.loop
-        ).result()
-
-    async def async_run(
-        self, variables: Optional[Sequence] = None, context: Optional[Context] = None
-    ) -> None:
-        """Run script.
 
-        This method is a coroutine.
-        """
-        self.last_triggered = date_util.utcnow()
-        if self._cur == -1:
-            self._log("Running script")
-            self._cur = 0
-
-        # Unregister callback if we were in a delay or wait but turn on is
-        # called again. In that case we just continue execution.
-        self._async_remove_listener()
-
-        for cur, action in islice(enumerate(self.sequence), self._cur, None):
+    async def _async_run(self, start):
+        for self._step, self._action in islice(
+            enumerate(self._script.sequence), start, None
+        ):
             try:
-                await self._handle_action(action, variables, context)
+                await getattr(self, f"_async_{_determine_action(self._action)}_step")()
             except _SuspendScript:
-                # Store next step to take and notify change listeners
-                self._cur = cur + 1
-                if self._change_listener:
-                    self.hass.async_add_job(self._change_listener)
-                return
+                raise
             except _StopScript:
                 break
-            except Exception:
-                # Store the step that had an exception
-                self._exception_step = cur
-                # Set script to not running
-                self._cur = -1
-                self.last_action = None
+            except Exception as err:
+                self._log_exception(err)
+                self._stop()
                 # Pass exception on.
                 raise
```

Comment:
Change listeners are not getting called now.

Reply:
That's the way it was. I was trying not to change "legacy" behavior. I suppose that since the exception it getting raised to the calling integration, it can do that itself it if wants (i.e., whatever the "change listener" function/method would have done that it passed into the script helper.) Is that ok, or do you want me to call the change listener before re-raising the exception?

Reply:
True. I think it's a bug. It seems like the state can get out of sync now ?

Reply:
I definitely looked at the current behavior as a bug, or at least a deficiency. I think it's cleaner, too, if the change listener is called no matter why the run has stopped. I'll change it.